core/services/gateway/handlers/vault/aggregator.go

```
@@ -0,0 +1,148 @@
+}
+
+func (a *baseAggregator) Aggregate(ctx context.Context, l logger.Logger, ar *activeRequest, currResp *jsonrpc.Response[json.RawMessage]) (*jsonrpc.Response[json.RawMessage], error) {
+	don, err := a.donForVaultCapability(ctx)
+	if err != nil {
+		return nil, fmt.Errorf("failed to get DON for vault capability: %w", err)
+	}
+
+	currResp, err = a.validateUsingSignatures(don.DON, don.Nodes, currResp)
+	if err == nil {
+		return currResp, nil
+	}
+
+	l.Debugw("failed to validate signatures, falling back to quorum aggregation", "error", err)
+	currResp, err = a.validateUsingQuorum(don.DON, ar, l)
+	if err != nil {
+		return nil, fmt.Errorf("failed to validate using quorum: %w", err)
+	}
+
+	return currResp, nil
+}
+
+func (a *baseAggregator) donForVaultCapability(ctx context.Context) (*capabilities.DONWithNodes, error) {
+	dons, err := a.capabilitiesRegistry.DONsForCapability(ctx, vaultcommon.CapabilityID)
+	if err != nil {
+		return nil, err
+	}
+	// TODO: Support multiple vault capabilities in the capability registry.
+	// For the initial Smartcon deployment there will be exactly one Vault capability
+	// split across both DON families.
+	if len(dons) != 1 {
+		return nil, fmt.Errorf("expected exactly one DON for vault capability, found %d", len(dons))
+	}
+
+	don := dons[0]
+	return &don, nil
+}
+
+func (a *baseAggregator) validateUsingQuorum(don capabilities.DON, ar *activeRequest, l logger.Logger) (*jsonrpc.Response[json.RawMessage], error) {
+	requiredQuorum := int(2*don.F + 1)
+
+	ar.mu.Lock()
+	defer ar.mu.Unlock()
+	if len(ar.responses) < requiredQuorum {
+		return nil, errInsufficientResponsesForQuorum
+	}
+
+	shaToCount := map[string]int{}
+	for _, r := range ar.responses {
+		sha, err := a.sha(r)
+		if err != nil {
+			l.Errorw("failed to compute digest of response during quorum validation, skipping...", "error", err)
+			continue
+		}
+		shaToCount[sha]++
+		if shaToCount[sha] >= requiredQuorum {
+			return r, nil
+		}
+	}
+
```

We likely need to fail faster here and handle that case.
If maximum shaToCount value + remaining responses is still less than requiredQuorum, we need to give an error indicating this request fails quorum.

As I believe, right now, if we don't get a quorum even after all responses are available, we don't respond back to user right?
Does the user see a timeout currently?